strawberry/subscriptions/protocols/graphql_transport_ws/handlers.py

```diff
@@ -243,106 +244,66 @@ async def handle_subscribe(self, message: SubscribeMessage) -> None:
-                    operation_name=message.payload.operationName,
-                )
 
-            result_source = get_result_source()
+        else:
+            result_source = self.schema.execute(
+                query=message.payload.query,
+                variable_values=message.payload.variables,
+                context_value=context,
+                root_value=root_value,
+                operation_name=message.payload.operationName,
+            )
 
         operation = Operation(self, message.id, operation_type)
 
-        # Handle initial validation errors
-        if isinstance(result_source, GraphQLExecutionResult):
-            assert operation_type == OperationType.SUBSCRIPTION
-            assert result_source.errors
-            payload = [err.formatted for err in result_source.errors]
-            await self.send_message(ErrorMessage(id=message.id, payload=payload))
-            self.schema.process_errors(result_source.errors)
-            return
-
         # Create task to handle this subscription, reserve the operation ID
         operation.task = asyncio.create_task(
             self.operation_task(result_source, operation)
         )
         self.operations[message.id] = operation
 
     async def operation_task(
-        self, result_source: AsyncGenerator, operation: Operation
+        self,
+        result_source: Awaitable[ExecutionResult] | Awaitable[SubscriptionResult],
+        operation: Operation,
     ) -> None:
         """The operation task's top level method. Cleans-up and de-registers the operation once it is done."""
         # TODO: Handle errors in this method using self.handle_task_exception()
         try:
-            await self.handle_async_results(result_source, operation)
-        except BaseException:  # pragma: no cover
-            # cleanup in case of something really unexpected
-            # wait for generator to be closed to ensure that any existing
-            # 'finally' statement is called
-            with suppress(RuntimeError):
-                await result_source.aclose()
-            if operation.id in self.operations:
-                del self.operations[operation.id]
+            first_res_or_agen = await result_source
+            # that's an immediate error we should end the operation
+            # without COMPLETE message
+            if isinstance(first_res_or_agen, PreExecutionError):
+                assert first_res_or_agen.errors
+                await operation.send_initial_errors(first_res_or_agen.errors)
+            # that's a mutation result
+            elif isinstance(first_res_or_agen, ExecutionResult):
+                await operation.send_next(first_res_or_agen)
+                await operation.send_message(CompleteMessage(id=operation.id))
+            else:
+                async for result in first_res_or_agen:
+                    await operation.send_next(result)
+                await operation.send_message(CompleteMessage(id=operation.id))
```

Comment:
I appreciate the increased readability! LGTM in regards to protocol-compliance too